src/components/Tags.tsx

```diff
@@ -20,122 +25,114 @@ const KeyCodes = {
   enter: 13,
 };
 
-type tagstruct = {
-  rank: number;
-  segment: string;
-  tag: { name: string };
-  _id: string;
-};
-
-interface MyProps {
-  tags: { id: string; text: string }[];
-  seg: any;
-  rank: number;
-  forceUpdate: any;
-}
-
-interface MyState {
-  tags: any;
-  suggestions: any;
-  k: number;
-}
-
 const delimiters = [KeyCodes.comma, KeyCodes.enter];
 
-class Tags extends React.Component<MyProps, MyState> {
-  constructor(props: MyProps) {
-    super(props);
-    this.state = {
-      tags: props.tags,
-      suggestions: suggestions,
-      k: 1,
+const Tags = ({
+  disabled,
+  currentSegment,
+  segmentIndex,
+  rank,
+  updateSegmentAt,
+  refresh,
+}: {
+  disabled: boolean;
+  currentSegment: Segment;
+  segmentIndex: number;
+  rank: number;
+  updateSegmentAt: (index: number, data: Partial<Segment>) => void;
+  refresh: [boolean];
+}): JSX.Element => {
+  const [reactTags, setReactTags] = React.useState([] as ReactTag[]);
+
+  const segmentTagToReactTag = (segTag: SegmentTag): ReactTag => {
+    return {
+      text: segTag.tag.name,
+      id: segTag.tag.name,
     };
+  };
 
-    this.handleDelete = this.handleDelete.bind(this);
-    this.handleAddition = this.handleAddition.bind(this);
-    this.handleDrag = this.handleDrag.bind(this);
-    this.handleTagClick = this.handleTagClick.bind(this);
-  }
-
-  handleDelete(i: any) {
-    const { tags } = this.state;
-    const delIndex = tags[i].id;
-
-    this.setState({
-      tags: tags.filter((tag: any, index: any) => index !== i),
-    });
-
-    //del from segment
-    for (let j = 0; j < this.props.seg.tags.length; j++) {
-      if (this.props.seg.tags[j].tag._id == delIndex) {
-        this.props.seg.tags.splice(j, 1);
-      }
+  const updateTags = (tags: ReactTag[]) => {
+    let currentSegmentTags = currentSegment.tags || [];
+    if (assertModelArrayType<SegmentTag>(currentSegmentTags, 'SegmentTag')) {
+      // Remove duplicates
+      tags = uniq(tags);
+      // Remove old tag if rank changed
+      currentSegmentTags = differenceBy(
+        currentSegmentTags,
+        tags.map(t => ({ tag: { name: t.text }, rank })),
+        'tag.name'
+      );
+      // Keep segments of other ranks
+      currentSegmentTags = currentSegmentTags.filter(t => t.rank !== rank);
+      // Add new tags to old
+      currentSegmentTags = [
+        ...currentSegmentTags,
+        ...tags.map(t => ({ tag: { name: t.text }, rank })),
+      ];
+      setReactTags(currentSegmentTags.filter(t => t.rank === rank).map(segmentTagToReactTag));
```

Comment:
Instead of using the general `this.setState` function for a class component, we can use the more concise `setReactTags` and only update the state that needs updating.